packages/engine-core/src/data-proxy/DataProxyEngine.ts

```diff
@@ -153,76 +167,119 @@ export class DataProxyEngine extends Engine {
       isolationLevel: transaction?.isolationLevel,
     }
 
-    const { batchResult } = await this.requestInternal<T>(body, headers, attempt)
+    const { batchResult } = await this.requestInternal<T, true>(body, headers)
 
+    // TODO: add elapsed to each result similar to BinaryEngine
+    // also check that the error handling is correct for batch
     return batchResult
   }
 
-  private async requestInternal<T>(body: Record<string, any>, headers: Record<string, string>, attempt: number) {
-    try {
-      this.logEmitter.emit('info', {
-        message: `Calling ${await this.url('graphql')} (n=${attempt})`,
-      })
+  private async requestInternal<T, Batch extends boolean = false>(
+    body: Record<string, any>,
+    headers: QueryEngineRequestHeaders,
+    itx?: InteractiveTransactionOptions<DataProxyTxInfoPayload>,
+  ): Promise<Batch extends true ? { batchResult: QueryEngineResult<T>[] } : QueryEngineResult<T>> {
+    return await this.withRetry({
+      actionGerund: 'querying',
+      callback: async ({ logHttpCall }) => {
+        const url = itx ? `${itx.payload.endpoint}/graphql` : await this.url('graphql')
+
+        logHttpCall(url)
+
+        const response = await request(url, {
+          method: 'POST',
+          headers: { ...runtimeHeadersToHttpHeaders(headers), ...this.headers },
+          body: JSON.stringify(body),
+          clientVersion: this.clientVersion,
+        })
 
-      const response = await request(await this.url('graphql'), {
-        method: 'POST',
-        headers: { ...headers, ...this.headers },
-        body: JSON.stringify(body),
-        clientVersion: this.clientVersion,
-      })
+        if (!response.ok) {
+          debug('graphql response status', response.status)
+        }
 
-      if (!response.ok) {
-        debug('graphql response status', response.status)
-      }
+        const e = await responseToError(response, this.clientVersion)
+        await this.handleError(e)
 
-      const e = await responseToError(response, this.clientVersion)
+        const data = await response.json()
 
-      if (e instanceof SchemaMissingError) {
-        await this.uploadSchema()
-        throw new ForcedRetryError({
-          clientVersion: this.clientVersion,
-          cause: e,
-        })
-      }
+        // TODO: headers contain `x-elapsed` and it needs to be returned
```

Comment:
this is out of scope of this pr and not related to itx — this is just something i noticed